crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2635,6 +2635,212 @@ impl World {
+            ));
+        };
+
+        // SAFETY:
+        // - `comp_ptr` is guaranteed to point to an object of type `type_id`
+        // - `reflect_from_ptr` was constructed for type `type_id`
+        // - Assertion that checks this equality is present
+        unsafe {
+            assert_eq!(
+                reflect_from_ptr.type_id(),
+                type_id,
+                "Mismatch between Ptr's type_id and ReflectFromPtr's type_id",
+            );
+
+            Ok(reflect_from_ptr.as_reflect(comp_ptr))
+        }
+    }
+
+    /// Retrieves a mutable reference to the given `entity`'s [`Component`] of the given `type_id` using
+    /// reflection.
+    ///
+    /// Requires [`#[derive(Reflect)`](derive@bevy_reflect::Reflect) on the [`Component`] and
+    /// `app.register_type::<TheComponent>()` to have been called.
+    ///
+    /// This is the mutable version of [`World::get_reflect()`], see its docs for more information
+    /// and an example.
+    ///
+    /// Just calling this method does not trigger [change detection](crate::change_detection).
+    ///
+    /// # Errors
+    ///
+    /// See [`GetComponentReflectError`](error::GetComponentReflectError) for the possible errors and their descriptions.
+    ///
+    /// # Example
+    ///
+    /// See the documentation for [`World::get_reflect()`].
+    ///
+    /// # Note
+    /// Requires the feature `bevy_reflect` (included in the default features).
+    #[inline]
+    #[cfg(feature = "bevy_reflect")]
+    pub fn get_reflect_mut(
+        &mut self,
+        entity: Entity,
+        type_id: TypeId,
+    ) -> Result<Mut<'_, dyn bevy_reflect::Reflect>, error::GetComponentReflectError> {
+        use bevy_reflect::ReflectFromPtr;
+
+        use crate::prelude::AppTypeRegistry;
+
+        use error::GetComponentReflectError;
+
+        // little clone() + read() dance so we a) don't keep a borrow of `self` and b) don't drop a
+        // temporary (from read()) too  early.
+        let Some(app_type_registry) = self.get_resource::<AppTypeRegistry>().cloned() else {
+            return Err(GetComponentReflectError::MissingAppTypeRegistry);
+        };
+        let type_registry = app_type_registry.read();
+
+        let Some(reflect_from_ptr) = type_registry.get_type_data::<ReflectFromPtr>(type_id) else {
+            return Err(GetComponentReflectError::MissingReflectFromPtrTypeData(
+                type_id,
+            ));
+        };
+
+        let Some(component_id) = self.components().get_id(type_id) else {
+            return Err(GetComponentReflectError::NoCorrespondingComponentId(
+                type_id,
+            ));
+        };
+
+        // HACK: Only required for the `None`-case/`else`-branch, but it borrows `self`, which will
+        // already be borrowed by `self.get_mut_by_id()`, and I didn't find a way around it.
+        let component_name = self
+            .components()
+            .get_name(component_id)
+            .map(ToString::to_string);
```

Comment:
I don't really want to keep this allocation on the happy-path even though it's only required in an error-case.
Just couldn't figure out how to prevent the double-borrow of `self`.

There is probably something `unsafe` that can be done?

Reply:
It borrows `self` immutably, but in the error branch `self` is already mutably borrowed from the call to `self.get_mut_by_id()`. Even though the error branch is the `None`-case, the borrow checker (apparently) counts the mutable borrow as still being live. I'd assume this is one of the current limitations of the borrow checker.

In `get_reflect()`, both are immutable borrows, which don't conflict, so it's fine.

Reply:
oh it's a [polonius](https://blog.rust-lang.org/inside-rust/2023/10/06/polonius-update.html) thing. we'll be able to remove eventually (or soon with `unsafe`) but i'd leave it as is for this PR.